### PR TITLE
Piot

### PR DESCRIPTION
**Pull request by vincentdass**
Raw change -

### CODE REVIEW

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
+import msRestAzure = require('./azure-arm-common');
+import webClient = require('./webClient');
+import azureServiceClient = require('./AzureServiceClient');
+import tl = require('vsts-task-lib/task');
+import Q = require('q');
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: can we follow the import * as pattern everywhere.

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+var parseString = require('xml2js').parseString;
+
+export class AzureAppService {
+    private appServiceName: string;
+    private resourceGroupName: string;
+    private slotName: string;
+    private endpoint: any;
+    private azureServiceClient: azureServiceClient.ServiceClient; 
+    private credentials: msRestAzure.ApplicationTokenCredentials;
+    private appDetails: Promise<any>;
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: start the private variables with underscore.

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import tl = require('vsts-task-lib/task');
+import Q = require('q');
+import * as querystring from "querystring";
+var parseString = require('xml2js').parseString;
+
+export class AzureAppService {
+    private appServiceName: string;
+    private resourceGroupName: string;
+    private slotName: string;
+    private endpoint: any;
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: is it possible to include an interface for endpoint as part of this PR ?

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 1 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: sure.

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    private azureServiceClient: azureServiceClient.ServiceClient; 
+    private credentials: msRestAzure.ApplicationTokenCredentials;
+    private appDetails: Promise<any>;
+
+    constructor(endpoint: any, name: string, resourceGroup: string, slot: string) {
+
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: should slot be a mandatory input ?

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 2 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: yes. will change it. thanks

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        this.credentials = new msRestAzure.ApplicationTokenCredentials(endpoint["servicePrincipalClientID"],
+            endpoint["tenantID"], endpoint["servicePrincipalKey"], endpoint["url"], 
+            endpoint["envAuthUrl"], endpoint["activeDirectoryResourceId"], endpoint["environment"].toLowerCase() == 'azurestack'
+        );
+        this.azureServiceClient = new azureServiceClient.ServiceClient(this.credentials, endpoint["subscriptionId"]);
+        this.slotName = (slot && slot.toLowerCase() == 'production') ? null : slot;
+    }
+
+    public async getResourceGroupName() {
+        var dataDeferred = Q.defer<string>();
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: use let instead of var.

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        if(this.resourceGroupName) {
+            console.log("getResourceGroupName in cache");
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: its better to check if(!!this.resourceGroupName).

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        return dataDeferred.promise;
+    }
+
+    // Move to azure-arm-common - Azure Monitor tasks
+    private async getResourceID() {
+        var dataDeferred = Q.defer<any>();
+        var httpRequest = new webClient.WebRequest();
+        httpRequest.method = 'GET';
+        httpRequest.uri = this.azureServiceClient.getRequestUri('//subscriptions/{subscriptionId}/resources', null,
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: two slashes before subscriptions in uri ?

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            }, (error) => {
+                dataDeferred.reject("Failed for get App Details: " + error.toString())
+            });
+        }
+
+        return dataDeferred.promise;
+    }
+
+    // Move to azure-arm-common - Azure Monitor tasks
+    private async getResourceID() {
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: if the resource group name is available, do we still need to make the rest call to fetch the resource id ?

File: Tasks/Common/azure-arm-rest/azure-app-service.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+            dataDeferred.resolve(this.appDetails);
+        }
+        else {
+            console.log(" getAppDetails not in cache");
+            var resourceGroupName: string = await this.getResourceGroupName();
+            var slotUrl: string = this.slotName ? `/slots/${this.slotName}` : '';
+            var httpRequest = new webClient.WebRequest();
+            httpRequest.method = 'GET';
+            httpRequest.uri = this.azureServiceClient.getRequestUri('//subscriptions/{subscriptionId}/resourceGroups/{ResourceGroupName}/providers/Microsoft.Web/sites/{AppServiceName}/{slotUrl}',
+                {
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-app-service.ts]: i think using `` will serve the same purpose as using getRequestUri.

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 3 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: yes. working on the same

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 4 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: no. will change it to optional

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 9 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: This function also removes additional slashes and update the URI with subscription ID.
and also encodes the URI

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 8 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: We save the RG name in a private variable and use the same

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 7 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: following the same pattern used by RG team. Anyway, the forward slashes are removed in the getRequestUri function

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 6 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: sure

File: Tasks/Common/azure-arm-rest/azure-app-service.ts (hunk 5 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-app-service.ts]: sure

**Pull request closed without merging**